### PR TITLE
Add support for migrating Table ACL for SQL Warehouse cluster in AWS using Instance Profile and Azure using SPN

### LINKED ISSUE #2238: [FEATURE]: By mapping SQL Warehourse Instance Profiles permission to UC permission when migrating tables
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

When using Glue as an External metastore, it is possible to set permissions at the Table level and also at the storage level.

Currently, instance profile permission of interactive clusters is migrated when doing table migration, but it doesn't cover instance profile used by SQL warehouse. if the warehouse has an instance profile that is different from the interactive cluster profile, then its permission will be missed. 

### Proposed Solution

Include SQL warehouse instance profile and its permission along with interactive cluster permission when migrating tables

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by HariGS-DB**
his PR contains the following changes for migrating Table ACL for SQL warehouse cluster used through Instance Profile in AWS scenario and spn info for Azure

Identifies all the SQL Warehouse
identifies all the instance profiles used by the data access config for aws and spn info for azure
identify the list of principals who have access to the warehouse (user, group, spn)
get the prefix these instance role arn or spn have access and the type of permission 
get the list of external locations matching the list of prefix
get the list of tables having the matching external location prefix from the table mapping CSV

Resolves #2238 

### Functionality

- [ ] added relevant user documentation
- [ ] added new CLI command
- [ ] modified existing command: `databricks labs ucx ...`
- [ ] added a new workflow
- [ ] modified existing workflow: `...`
- [ ] added a new table
- [ ] modified existing table: `...`

### Tests

- [ ] manually tested
- [ ] added unit tests
- [ ] added integration tests
- [ ] verified on staging environment (screenshot attached)

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 1]
```diff
@@ -403,14 +405,27 @@ def _get_cluster_to_instance_profile_mapping(self) -> dict[str, str]:
...
+    def _update_warehouse_to_instance_profile_mapping(
+        self,
+    ) -> dict[str, str]:
+        warehouse_instance_profiles = {}
+        sql_config = self._ws.warehouses.get_workspace_warehouse_config()
+        if sql_config.instance_profile_arn is not None:
+            role_name = sql_config.instance_profile_arn
+            for warehouse in self._ws.warehouses.list():
+                if warehouse.id is not None:
+                    warehouse_instance_profiles[warehouse.id] = role_name
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:417]: can you add the integration test for AWS and attach a screenshot of it passing?

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 2]
```diff
@@ -628,11 +659,11 @@ def _get_grants(
 
         return grants
 
-    def _get_cluster_principal_mapping(self, cluster_id: str) -> list[str]:
+    def _get_cluster_principal_mapping(self, cluster_id: str, object_type: str) -> list[str]:
         # gets all the users,groups,spn which have access to the clusters and returns a dataclass of that mapping
         principal_list = []
         try:
-            cluster_permission = self._ws.permissions.get("clusters", cluster_id)
+            cluster_permission = self._ws.permissions.get(object_type, cluster_id)
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/grants.py:666]: nice

**Review by nfx — COMMENTED** [review]
nice!

**Comment by HariGS-DB** [comment]
image

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 1 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:417]: added aws int test.

Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:417]: could not test azure getting an issue when trying to set SQL data access configuration 
https://github.com/databricks/databricks-sdk-py/issues/305

File: src/databricks/labs/ucx/assessment/azure.py [hunk 3]
```diff
@@ -193,3 +193,14 @@ def get_cluster_to_storage_mapping(self):
                 set_service_principals = self._get_azure_spn_from_cluster_config(cluster)
                 spn_cluster_mapping.append(ServicePrincipalClusterMapping(cluster.cluster_id, set_service_principals))
         return spn_cluster_mapping
+
+    def get_warehouse_to_storage_mapping(self):
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/assessment/azure.py:197]: Add type hint

File: src/databricks/labs/ucx/assessment/azure.py [hunk 4]
```diff
@@ -193,3 +193,14 @@ def get_cluster_to_storage_mapping(self):
...
+        # this function gives a mapping between a sql warehouse and the spn used by it
+        set_service_principals = set[AzureServicePrincipalInfo]()
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/assessment/azure.py:199]: looks redundant

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 5]
```diff
@@ -434,12 +449,23 @@ def get_eligible_locations_principals(self) -> dict[str, dict]:
...
+            eligible_locations.update(
+                self._get_external_locations(role_compute, external_locations, permission_mappings)
+            )
             if len(eligible_locations) == 0:
                 continue
-            cluster_locations[cluster_id] = eligible_locations
-        return cluster_locations
+            compute_locations.append(ComputeLocations(cluster_id, eligible_locations, "clusters"))
+        for warehouse_id, role_compute in warehouse_instance_profiles.items():
+            eligible_locations = {}
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/hive_metastore/grants.py:461]: Why create an empty dictionairy before the update?

File: src/databricks/labs/ucx/hive_metastore/grants.py [hunk 6]
```diff
@@ -475,14 +501,15 @@ def __init__(
         self._spn_crawler = spn_crawler
         self._installation = installation
 
-    def get_eligible_locations_principals(self) -> dict[str, dict]:
-        cluster_locations = {}
-        eligible_locations = {}
+    def get_eligible_locations_principals(self) -> list[ComputeLocations]:
+        compute_locations = []
+        eligible_locations: dict[str, str] = {}
```
Comment (JCZuurmond) [inline src/databricks/labs/ucx/hive_metastore/grants.py:506]: Looks redundant

File: tests/integration/hive_metastore/test_migrate.py [hunk 7]
```diff
@@ -564,6 +563,33 @@ def test_migrate_external_tables_with_principal_acl_aws(
...
+    ws, make_user, prepared_principal_acl, make_warehouse_permissions, make_warehouse, env_or_skip
+):
+    if not ws.config.is_aws:
+        pytest.skip("temporary: only works in aws test env")
+    ctx, table_full_name, _, _ = prepared_principal_acl
+    ctx.with_dummy_resource_permission()
+    warehouse = make_warehouse()
+    table_migrate = ctx.tables_migrator
+    user = make_user()
+    make_warehouse_permissions(
```
Comment (JCZuurmond) [inline tests/integration/hive_metastore/test_migrate.py:577]: Do I understand correctly that we are migrating user access to locations to which it has access via access to a warehouse? Where do we mount or connect locations to a warehouse?

**Review by JCZuurmond — COMMENTED** [review]
Mainly have some clarification questions

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 5 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py:461]: You could get 2 sp details from one cluster and 3 spn from second cluster, hence I am resetting the dict in the outer loop before the cluster loops to next cluster

File: tests/integration/hive_metastore/test_migrate.py (hunk 7 above)
Comment (HariGS-DB) [reply tests/integration/hive_metastore/test_migrate.py:577]: Yes that is right. we dont need to connect location to the warehouse, that's done at the UC level. So if a user has access to a table in non uc (through the spn configured in the warehouse), that user will have access to the location and the table once it is migrated to UC, warehouse spn is not used in UC world

File: src/databricks/labs/ucx/assessment/azure.py (hunk 4 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/assessment/azure.py]: removed

File: src/databricks/labs/ucx/assessment/azure.py (hunk 3 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/assessment/azure.py]: added

File: src/databricks/labs/ucx/hive_metastore/grants.py (hunk 6 above)
Comment (HariGS-DB) [reply src/databricks/labs/ucx/hive_metastore/grants.py]: removed

**Review by nfx — CHANGES_REQUESTED** [review]

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**